src/tests/JIT/IL_Conformance/Old/Base/add_.il

```diff
@@ -17,12 +17,12 @@
 //-----------------------------------------------------
 
 
-.assembly 'add'{
+.assembly 'add_'{
```

Comment:
I think something may have gone wrong in the dedup renaming.  I don't think this and others need to be renamed at the project/assembly level, which will add a lot of churn to the test source history.  Only the namespace should need the rewrite.

(old\base\add.il is the only add.il, but old\base\add.il, old\conformance_base\add_i.il, and others all have namespace 'add')

Reply:
IIRC the only limitations are that the assembly name must match the project name and that assembly names must be unique within a single merged test wrapper, otherwise we cannot generate functional C# code calling their entrypoints - in the former case (assembly vs. project name mismatch) the Roslyn generator at some point thinks that the project (dll) name is the assembly name and uses that for module qualification of the method call which subsequently fails at runtime because runtime compares it with the logical assembly name. In the latter case (assembly names unique within a test wrapper) this is due to the fact that all the test assemblies are ultimately copied to the merged test wrapper output folder so same-named assemblies would stomp on each other; this is "silent" in the sense that nothing fails but you'll see a lower number of tests compared to the non-merged version. A lower number of tests executed by <code>src\tests\run</code> can also indicate premature termination of the merged test wrapper that happened to me for several JIT/Methodical tests that were using <code>Environment.Exit</code> to report success.

Reply:
This is the most annoying for silly recursive tests that need to exit early, in a handful of cases I had to do non-trivial rewrites for this reason. IIRC I typically introduced a static variable representing the exit code, changed the code calling Environment.Exit to set that instead, patched the recursive descents to check the variable prior to the call and the main method to use the variable for the final return value. There were less than half a dozen cases like these so I guess each of them is unique, feel free to devise your own method to overcome the Environment.Exit problem.

Reply:
You are right, I was matching namespaces with il files, but it is no need.

Reply:
I'm pretty sure that there is something required at the namespace level (or more specifically that type names need to be unique), but I'm not sure exactly what the restriction is.  My original understanding (I think from ILTransform code) was that this applied to the type containing the test entrypoint (perhaps due to invoking it without assembly scoping?).  However, I then hit build errors for types that weren't the container for test entrypoints, so I adjusted ILTransform to look at all of the type names.

It is quite possible that I misinterpreted what was going on.

That said, deduplicating the types will be necessary if we try to combine multiple tests into one C# compiler invocation with many files (as a future consolidation step to save time).  Perhaps this could be relevant on the IL side though the tool invocation might be trickier.

We have ILTransform scanning for Environment.Exit, adding RequiresProcessIsolation for them, and reporting them in the log.  I think we've been pretty lucky that they are rare / non-existent across jit\regression / il_conformance, but we will likely hit this again.

Reply:
Sounds great. IIRC @jkoritzinsky was looking into that at some point; there's some hacky partially supported form of syntax that lets you address the same-typed name in multiple assemblies via some obscure form of qualification, however @jkotas suggested it would be better to clean these cases up so that's what I did, typically in the Loader/classloader/TypeGeneratorTests all tests were using the same namespace and class name i.o.w. the fully qualified entrypoint name was the same for all of them, I believe that was the point of this particular deduplication.